jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/sender/JMSClientConnector.java

```diff
@@ -77,76 +77,141 @@ public boolean send(CarbonMessage carbonMessage, CarbonCallback carbonCallback)
-                MapMessage mapMessage = (MapMessage) message;
-                Enumeration<String> mapNames = ((MapCarbonMessage) carbonMessage).getMapNames();
-                while (mapNames.hasMoreElements()) {
-                    String key = mapNames.nextElement();
-                    mapMessage.setString(key, ((MapCarbonMessage) carbonMessage).getValue(key));
-                }
-            }
+            setupMessageProducer(propertyMap.entrySet());
 
-            if (carbonMessage.getProperty(JMSConstants.PERSISTENCE) != null &&
-                carbonMessage.getProperty(JMSConstants.PERSISTENCE).equals(false)) {
-                messageProducer.setDeliveryMode(DeliveryMode.NON_PERSISTENT);
-            }
+            Message jmsMessage = createJmsMessage(carbonMessage, propertyMap);
 
-            if (message != null) {
-                messageProducer.send(message);
-            } else {
-              throw new ClientConnectorException("Exception occured while creating the message");
-            }
+            sendMessage(carbonMessage, jmsMessage);
 
-        } catch (JMSException e) {
-            throw new ClientConnectorException("Exception occurred while sending the message", e);
         } finally {
             if (jmsConnectionFactory != null) {
                 try {
                     jmsConnectionFactory.closeMessageProducer(messageProducer);
                     jmsConnectionFactory.closeSession(session);
                     jmsConnectionFactory.closeConnection(connection);
                 } catch (JMSConnectorException e) {
-                    logger.error("Exception occured when closing connection " + e.getMessage(), e);
+                    logger.error("Exception occurred when closing connection. Error: " + e.getMessage(), e);
                 }
             }
         }
         return false;
     }
 
+    private void sendMessage(CarbonMessage carbonMessage, Message jmsMessage) throws ClientConnectorException {
+        int deliveryMode = DeliveryMode.PERSISTENT;
+        if (getHeader(carbonMessage, JMSConstants.JMS_DELIVERY_MODE).
+                equalsIgnoreCase(JMSConstants.NON_PERSISTENT_DELIVERY_MODE)) {
+            deliveryMode = DeliveryMode.NON_PERSISTENT;
+        }
+        int priority = JMSConstants.DEFAULT_PRIORITY;
+        String value = carbonMessage.getHeader(JMSConstants.JMS_PRIORITY);
+        if (value != null) {
+            priority = Integer.parseInt(value);
+        }
+
+        long ttl = 0;
+        value = carbonMessage.getHeader(JMSConstants.JMS_EXPIRATION);
+        if (value != null) {
+            ttl = Long.parseLong(value);
+        }
+        try {
+            messageProducer.send(jmsMessage, deliveryMode, priority, ttl);
+        } catch (JMSException e) {
+            throw new ClientConnectorException("Error occurred while sending the message to JMS provider. " +
+                                                e.getMessage(), e);
```

Comment:
Add the delivery mode priority details to the error message. Kept the e.getMessage since we need to propagate the root cause to the Ballerina programme until we come up with an exception mapping implementation